HPCC-13381 Add #option ('obfuscateOutput') to hide original ECL

ghalliday opened the pull request:
Useful if you wish to publish a query to a third party, but hide details
of the ECL implementation.

Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

ghalliday committed:
HPCC-13381 Add #option ('obfuscateOutput') to hide original ECL

Useful if you wish to publish a query to a third party, but hide details
of the ECL implementation.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

richardkchapman commented:
Typo (part should say party) in commit message.

ghalliday commented:
@richardkchapman Fixed commit message

richardkchapman commented:
Have you tested that platforms can still run ok with this setting enabled?

jamienoss commented:
@ghalliday is this different from the original PR? I don't remember the comparison being so 'jumbled' and difficult to follow, perhaps github has made a hash of the comp.?

richardkchapman commented:
Compare with ?w=0 and it's a lot easier to follow

jamienoss commented:
@richardkchapman bingo, cheers. I had forgotten that Gavin had suggested this in the original PR. Thanks again.

jamienoss commented:
@ghalliday, @richardkchapman the code looks good to me.

I still think we should add 'showEclInGraph' to the usage rather than deferring it to https://track.hpccsystems.com/browse/HPCC-12313.

Remind me, are all -f options accessible as #options in the actual ECL? If so, this change should contain at least one regression example.

richardkchapman commented:
I agree it's worth adding an example to the runtime regression suite that includes the obfuscate option to ensure that such code can still run.

Regarding documenting all options, I think better to leave the obscure options out until HPCC-12313 is done (which can then include support for levels of detail...)

AttilaVamos commented:
Automated Smoketest Build: success

ghalliday committed:
HPCC-13381 Add #option ('obfuscateOutput') to regression suite example

Also, make sure that the ECL source and archive is not included in the output.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

AttilaVamos commented:
Automated Smoketest Build: success

jamienoss commented on testing/regress/ecl/aggidx1.ecl:
```diff
@@ -24,6 +24,8 @@ useSequential := #IFDEFINED(root.useSequential, false);
 
 //--- end of version configuration ---
 
+#option ('obfuscateOutput', true);
```
I don't think we should be reusing tests. aggidx1.ecl should be copied and then the new code added.

richardkchapman replied:
We have test suite that takes so long to run that no-one runs it, because we have a tendency to add a new test just to be sure rather than using coverage analysis to ensure that a particular bit of code is tested. If we take your suggestion to its logical conclusion we would have one test per closed Jira? one test per source code function? 

In general it's good (for working out what has gone wrong) if one test covers one bit of functionality, but we have to balance that against the need to have a test suite that people actually run.

ghalliday replied:
In this case I wanted to use a test which used indexes since that would be something that might be broken by removing meta information from the graph.  If I had added a new test it would have essentially been a clone of one of the existing ones - so wouldn't have had any benefit.

If it is testing a new piece of functionality then I agree.  This isn't though.

jamienoss replied:
I was waiting for this line of reasoning.
Having  a cap on the size of our regression suite (note no plural) makes no sense unless you also want to cap all further production of the src *and* are content that the suite is already complete.
Perhaps we need another and more thorough suite to better tailor their usage.

If we start changing existing tests then we need to be 100% sure that we haven't negated *any* factor of its original purpose.

As you pointed out, and I concur, a runtime test for this fix should be a must, given it's nature, and should be worthy of a new test unto itself. 

Having a individual test per closed jira is almost required until we correctly utilise coverage analysis, such that we can state that an existing test covers that which is required to test (and future test) the given code change.

In this particular case, perhaps I should have said that "aggidx1.ecl should *at least* be copied...". Any other PR that adds a test is subject to the review process, such that aggidx1.ecl now is. Not that there is anything wrong with this by default but there maybe be if this is not as complete as one might expect.

richardkchapman replied:
Sorry, but we have to be pragmatic. Making the regression suite bigger just because it feels right means making the regression suite less likely to be run. In the absence of coverage metrics we have to rely on common sense to know when a test is adding something and when it is not. We have 12000 closed Jiras - there are 500ish tests in the regression suite...

ghalliday commented:
@jamienoss can you review the new commit as well.  There is a pull request in the LN repository:  https://github.com/hpcc-systems/LN/pull/736

jamienoss commented:
@ghalliday, I was doing just that and about to comment on
void CLocalWUQuery::removeAssociatedFile
not being used, but perhaps this is in LN.

ghalliday commented:
@jamienoss correct

jamienoss commented:
@ghalliday, @richardkchapman, the new commit looks to good to me also.
